chore(gatsby): convert inference-metadata to typescript

Kornil opened the pull request:
Convert  `inference-metadata` reducer to TypeScript.

## Related Issues

https://github.com/gatsbyjs/gatsby/issues/21995

Kornil commented on packages/gatsby/src/redux/reducers/inference-metadata.ts:
```diff
@@ -123,8 +82,8 @@ const incrementalReducer = (state = {}, action) => {
-      state[type] = deleteNode(state[type], { fields: previousFields })
-      state[type] = addNode(state[type], { fields: node.fields })
+      state[type] = deleteNode(state[type], { ...node, fields: previousFields })
+      state[type] = addNode(state[type], { ...node, fields: node.fields })
```
These 2 changes have the potential to affect production code as we are now sending a full node instead of just fields. I based this on the fact that the action themselves (`deleteNode` and `addNode`) are already typed and coded to accept a full node as second parameter.

blainekasten replied:
This feels _correct_, so it's just a matter of if tests will pass correctly with this change. Thanks for figuring this out!

Kornil replied:
Tests seem to pass nicely so it should be all good, I just wanted to make sure this change was seen.

blainekasten commented on packages/gatsby/src/redux/reducers/inference-metadata.ts:
```diff
@@ -161,3 +120,55 @@ const incrementalReducer = (state = {}, action) => {
+const StepsEnum = {
+  initialBuild: `initialBuild`,
+  incrementalBuild: `incrementalBuild`,
+}
```
This should be a TS enum

blainekasten commented on packages/gatsby/src/redux/reducers/inference-metadata.ts:
```diff
@@ -161,3 +120,55 @@ const incrementalReducer = (state = {}, action) => {
+  }
+}
+
+module.exports = (
```
this should be a named export and update the import reference.

e.g.
```js
export function inferenceMetadataReducer(state, action) { .... }
```

Kornil commented:
Thank you for the review @blainekasten, I've addressed your feedback, feel free to resolve the conversations if you are ok with the new code :heavy_check_mark:
